### ISSUE TITLE
`_ghost` error when checking the translated proof

### ISSUE THREAD

**Issue by MartyO256** [Beluga-lang/Beluga#229]
Load the following signature:
```
LF kind : type = t : kind | sing : con -> kind
and con : type =;

LF cn-of : con -> kind -> type =;

LF kd-wf : kind -> type =  kd-wf/sing : cn-of C t -> kd-wf (sing C);

LF kd-equiv : kind -> kind -> type =
| kd-equiv/refl : kd-wf K -> kd-equiv K K

| kd-equiv/symm : kd-equiv K1 K2 -> kd-equiv K2 K1

| kd-equiv/trans : kd-equiv K1 K2 -> kd-equiv K2 K3 -> kd-equiv K1 K3

| kd-equiv/sing : cn-equiv C1 C2 t -> kd-equiv (sing C1) (sing C2)

and cn-equiv : con -> con -> kind -> type =
| cn-equiv/refl : cn-of C K -> cn-equiv C C K

| cn-equiv/symm : cn-equiv C1 C2 K -> cn-equiv C2 C1 K

| cn-equiv/trans : cn-equiv C1 C2 K -> cn-equiv C2 C3 K -> cn-equiv C1 C3 K

| cn-equiv/sing : cn-equiv C C' t -> cn-equiv C C' (sing C)
;

LF kind-eq : kind -> kind -> type =
| kind-eq/i : kind-eq K K
;

schema conblock+conbind =
  % conblock
  block (a : con) +
  % conbind
  some [K : kind] block (a : con, d : cn-of a K);

proof kd-equiv-resp :
  (g : conblock+conbind)
  [g |- kind-eq K1 K1'] ->
  [g |- kind-eq K2 K2'] ->
  [g |- kd-equiv K1 K2] ->
    [g |- kd-equiv K1' K2'] =
/ total /
intros
{ g : conblock+conbind,
  K : (g |- kind),
  K1' : (g |- kind),
  K1 : (g |- kind),
  K2' : (g |- kind)
| eq : [g |- kind-eq K K1'],
  eq1 : [g |- kind-eq K1 K2'],
  equiv : [g |- kd-equiv K K1]
; split eq as
  case kind-eq/i:
  { g : conblock+conbind,
    K1' : (g |- kind),
    K1 : (g |- kind),
    K2' : (g |- kind)
  | eq : [g |- kind-eq K1' K1'],
    eq1 : [g |- kind-eq K1 K2'],
    equiv : [g |- kd-equiv K1' K1]
  ; split eq1 as
    case kind-eq/i:
    { g : conblock+conbind, K1' : (g |- kind), K2' : (g |- kind)
    | eq : [g |- kind-eq K1' K1'],
      eq1 : [g |- kind-eq K2' K2'],
      equiv : [g |- kd-equiv K1' K2']
    ; solve equiv
    }
  }
}
;

schema conbind = some [K : kind] block (a : con, d : cn-of a K);

LF kd-equiv-sing-form/e : kind -> type =
| kd-equiv-sing-form/i :
  { C' : con }
  kind-eq K (sing C') -> kd-equiv-sing-form/e K
;

proof kd-equiv-sing-form :
  (g : conbind)
  [g |- kd-equiv (sing C) K] ->
    [g |- kd-equiv-sing-form/e K] =
/ trust /
intros
{ g : conbind, C : (g |- con), K : (g |- kind)
| equiv : [g |- kd-equiv (sing C) K]
; split equiv as
  case kd-equiv/sing:
  { g : conbind,
    C : (g |- con),
    C2 : (g |- con),
    Dequiv : (g |- cn-equiv C C2 t)
  | equiv : [g |- kd-equiv (sing C) (sing C2)]
  ; solve [g |- kd-equiv-sing-form/i _ (kind-eq/i )]
  }
  case kd-equiv/trans:
  { g : conbind,
    C : (g |- con),
    K : (g |- kind),
    K2 : (g |- kind),
    Dequiv : (g |- kd-equiv (sing C) K2),
    Dequiv1 : (g |- kd-equiv K2 K)
  | equiv : [g |- kd-equiv (sing C) K]
  ; split kd-equiv-sing-form [g |- Dequiv] as
    case kd-equiv-sing-form/i:
    { g : conbind,
      C : (g |- con),
      K : (g |- kind),
      K2 : (g |- kind),
      Dequiv : (g |- kd-equiv (sing C) K2),
      Dequiv1 : (g |- kd-equiv K2 K),
      C1 : (g |- con),
      Deq : (g |- kind-eq K2 (sing C1))
    | equiv : [g |- kd-equiv (sing C) K]
    ; by kd-equiv-resp [g |- Deq] [g |- kind-eq/i ] [g |- Dequiv1]
      as Dequiv'' unboxed;
      split kd-equiv-sing-form [g |- Dequiv''] as
      case kd-equiv-sing-form/i:
      { g : conbind,
        C : (g |- con),
        K : (g |- kind),
        K2 : (g |- kind),
        Dequiv : (g |- kd-equiv (sing C) K2),
        Dequiv1 : (g |- kd-equiv K2 K),
        C1 : (g |- con),
        Deq : (g |- kind-eq K2 (sing C1)),
        Dequiv'' : (g |- kd-equiv (sing C1) K),
        C2 : (g |- con),
        Deq1 : (g |- kind-eq K (sing C2))
      | equiv : [g |- kd-equiv (sing C) K]
      ; solve [g |- kd-equiv-sing-form/i _ Deq1]
      }
    }
  }
  case kd-equiv/symm:
  { g : conbind,
    C : (g |- con),
    K : (g |- kind),
    Dequiv : (g |- kd-equiv K (sing C))
  | equiv : [g |- kd-equiv (sing C) K]
  ; split kd-equiv-sing-form' [g |- Dequiv] as
    case kd-equiv-sing-form/i:
    { g : conbind,
      C : (g |- con),
      K : (g |- kind),
      Dequiv : (g |- kd-equiv K (sing C)),
      C1 : (g |- con),
      Deq : (g |- kind-eq K (sing C1))
    | equiv : [g |- kd-equiv (sing C) K]
    ; solve [g |- kd-equiv-sing-form/i _ Deq]
    }
  }
  case kd-equiv/refl:
  { g : conbind, C : (g |- con), Dwf : (g |- kd-wf (sing C))
  | equiv : [g |- kd-equiv (sing C) (sing C)]
  ; solve [g |- kd-equiv-sing-form/i _ (kind-eq/i )]
  }
}

and proof kd-equiv-sing-form' :
  (g : conbind)
  [g |- kd-equiv K (sing C)] ->
    [g |- kd-equiv-sing-form/e K] =
/ trust /
intros
{ g : conbind, K : (g |- kind), C : (g |- con)
| equiv : [g |- kd-equiv K (sing C)]
; split equiv as
  case kd-equiv/sing:
  { g : conbind,
    C1 : (g |- con),
    C : (g |- con),
    Dequiv : (g |- cn-equiv C1 C t)
  | equiv : [g |- kd-equiv (sing C1) (sing C)]
  ; solve [g |- kd-equiv-sing-form/i _ (kind-eq/i )]
  }
  case kd-equiv/trans:
  { g : conbind,
    K : (g |- kind),
    C : (g |- con),
    K2 : (g |- kind),
    Dequiv : (g |- kd-equiv K K2),
    Dequiv1 : (g |- kd-equiv K2 (sing C))
  | equiv : [g |- kd-equiv K (sing C)]
  ; split kd-equiv-sing-form' [g |- Dequiv1] as
    case kd-equiv-sing-form/i:
    { g : conbind,
      K : (g |- kind),
      C : (g |- con),
      K2 : (g |- kind),
      Dequiv : (g |- kd-equiv K K2),
      Dequiv1 : (g |- kd-equiv K2 (sing C)),
      C1 : (g |- con),
      Deq : (g |- kind-eq K2 (sing C1))
    | equiv : [g |- kd-equiv K (sing C)]
    ; by kd-equiv-resp [g |- kind-eq/i ] [g |- Deq] [g |- Dequiv]
      as Dequiv'' unboxed;
      split kd-equiv-sing-form' [g |- Dequiv''] as
      case kd-equiv-sing-form/i:
      { g : conbind,
        K : (g |- kind),
        C : (g |- con),
        K2 : (g |- kind),
        Dequiv : (g |- kd-equiv K K2),
        Dequiv1 : (g |- kd-equiv K2 (sing C)),
        C1 : (g |- con),
        Deq : (g |- kind-eq K2 (sing C1)),
        Dequiv'' : (g |- kd-equiv K (sing C1)),
        C2 : (g |- con),
        Deq1 : (g |- kind-eq K (sing C2))
      | equiv : [g |- kd-equiv K (sing C)]
      ; solve [g |- kd-equiv-sing-form/i _ Deq1]
      }
    }
  }
  case kd-equiv/symm:
  { g : conbind,
    K : (g |- kind),
    C : (g |- con),
    Dequiv : (g |- kd-equiv (sing C) K)
  | equiv : [g |- kd-equiv K (sing C)]
  ; split kd-equiv-sing-form [g |- Dequiv] as
    case kd-equiv-sing-form/i:
    { g : conbind,
      K : (g |- kind),
      C : (g |- con),
      Dequiv : (g |- kd-equiv (sing C) K),
      C1 : (g |- con),
      Deq : (g |- kind-eq K (sing C1))
    | equiv : [g |- kd-equiv K (sing C)]
    ; solve [g |- kd-equiv-sing-form/i _ Deq]
    }
  }
  case kd-equiv/refl:
  { g : conbind, C : (g |- con), Dwf : (g |- kd-wf (sing C))
  | equiv : [g |- kd-equiv (sing C) (sing C)]
  ; solve [g |- kd-equiv-sing-form/i _ (kind-eq/i )]
  }
}
;

proof injective-sing :
  (g : conbind)
  [g |- kd-equiv (sing C) (sing C')] ->
    [g |- cn-equiv C C' t] =
/ trust /
?
;
```
Complete the proof of `injective-sing` using the following:
```
split x
invert kd-equiv-sing-form [_|-X3]
by kd-equiv-resp [_|-kind-eq/i] [_|-X7] [_|-X3] as D12' unboxed
by kd-equiv-resp [_|-X7] [_|-kind-eq/i] [_|-X4] as D23' unboxed
by injective-sing [_|-D12'] as D12'' unboxed
by injective-sing [_|-D23'] as D23'' unboxed
solve [_|-cn-equiv/trans D12'' D23'']
by injective-sing [_|-X2] as D' unboxed
solve [_|-cn-equiv/symm D']
msplit X1
solve [_|-cn-equiv/refl X3]
```
Here's the translation generated program:
```
Translation generated program:
  fn x2 =>
  case x2 of
  | [_ |- kd-equiv/sing X2] =>
    [_ |- X2]
  | [_ |- kd-equiv/trans X3 X4] =>
    let [_ |- kd-equiv-sing-form/i X6 X7] = kd-equiv-sing-form [_ |- X3] in
    let [_ |- D12'] = kd-equiv-resp [_ |- kind-eq/i ] [_ |- X7] [_ |- X3] in
    let [_ |- D23'] = kd-equiv-resp [_ |- X7] [_ |- kind-eq/i ] [_ |- X4] in
    let [_ |- D12''] = injective-sing [_ |- D12'] in
    let [_ |- D23''] = injective-sing [_ |- D23'] in
    [_ |- cn-equiv/trans D12'' D23'']
  | [_ |- kd-equiv/symm X2] =>
    let [_ |- D'] = injective-sing [_ |- X2] in [_ |- cn-equiv/symm D']
  | [_ |- kd-equiv/refl X1] =>
    let [_ |- kd-wf/sing X3] = [_ |- X1] in [_ |- cn-equiv/refl X3]
```
A `_ghost` error is thrown when checking the translated proofs:
```
No theorems left. Checking translated proofs.
- An error occurred when checking the translated proofs.
  Please report this as a bug.
  _ghost, line 1, column 0:
Type checking encountered ill-typed meta-object. This is a bug in type reconstruction.
    Expected context:
      _
    Given context:
      _
In expression: [_ |- X2]
```
This error is not thrown after reloading the signature.

**Comment by tsani** [comment]
Translations are generated at the moment that the proof is completed, and not when loading the signature. It could be worthwhile to change it so that we always generate and check a translation after checking a complete proof.

Anyway, this is a very strange issue. The context of `X2` should be correctly inferred, and I don't see any obvious issue with the translated program. One should try replacing the `proof` of `injective-sing` with a `rec` and using the program obtained from the translation. If it parses and checks fine that way, then this would suggest an "invisible" problem with the translated program and one would need to dig deeper into the translation.

**Comment by tsani** [comment]
Typechecking does succeed when we add the following to the end of the file.

```
rec injective-sing :
  (g : conbind)
  [g |- kd-equiv (sing C) (sing C')] ->
    [g |- cn-equiv C C' t] =
/ trust /
fn x2 =>
  case x2 of
  | [_ |- kd-equiv/sing X2] =>
    [_ |- X2]
  | [_ |- kd-equiv/trans X3 X4] =>
    let [_ |- kd-equiv-sing-form/i X6 X7] = kd-equiv-sing-form [_ |- X3] in
    let [_ |- D12'] = kd-equiv-resp [_ |- kind-eq/i ] [_ |- X7] [_ |- X3] in
    let [_ |- D23'] = kd-equiv-resp [_ |- X7] [_ |- kind-eq/i ] [_ |- X4] in
    let [_ |- D12''] = injective-sing [_ |- D12'] in
    let [_ |- D23''] = injective-sing [_ |- D23'] in
    [_ |- cn-equiv/trans D12'' D23'']
  | [_ |- kd-equiv/symm X2] =>
    let [_ |- D'] = injective-sing [_ |- X2] in [_ |- cn-equiv/symm D']
  | [_ |- kd-equiv/refl X1] =>
    let [_ |- kd-wf/sing X3] = [_ |- X1] in [_ |- cn-equiv/refl X3]
;
```

So there is some kind of invisible problem in the internal syntax that disappears when it is erased and printed out.

**Comment by tsani** [comment]
The problem seems to be an off-by-one error. After changing some prints to force showing implicit context variables, I see the following in the debug log.

```
[checkMetaObj] cD = g : conbind, C : (g |- con), C' : (g |- con),
                    X2 : (g |- cn-equiv C C' t)
               cPsi = g
               phat = C
```

but phat is supposed to be `g`.

**Comment by tsani** [comment]
It turns out that the issue was due to buggy automation code. `auto-solve-trivial` was not shifting the erased LF context bundled with the contextual term it constructed. Most of the time this isn't a problem (in particular during printing) because I added type annotations to all contextual objects at some point. So during printing, instead of using the erased LF context (with which it is impossible to generate type annotations for bound variables as required by type reconstruction; see #185) I used the annotation. This is why the problem seemed to go away after reloading: the pretty-printer would generate the right output *despite* the internal AST being a bit messed up.

The solution was to make `auto-solve-trivial` shift not only the annotation on the box, but also the erased LF context (`dctx_hat`).
